### PR TITLE
[WIP] Add sync collection and record GET verbs.

### PR DESCRIPTION
**Pull request by Natim**

### CODE REVIEW

**Comment by Natim** [comment]
I ahve some refactor and caching that should takes place with regards to the token server.

File: syncto/tests/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    return text_type(uuid4())
+
+
+def generate_made_up_uuid4():
+    uuid_bytes = list(uuid4().bytes)
+    uuid_bytes[6] = binary_type('\x40')
+    uuid_bytes[8] = binary_type('\x80')
+    uuid = UUID(hexlify(binary_type(''.join(uuid_bytes))))
+    return text_type(uuid)
+
```
Comment (Natim) [inline syncto/tests/test_utils.py:24]: Add also a list of edge cases that we must test.

File: syncto/tests/test_utils.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+class Bytes2UUID4Test(unittest.TestCase):
+    def test_a_wrong_uuid4_array_fails(self):
+        try:
+            bytes_to_uuid4(os.urandom(16))
+        except ValueError:
+            pass
+        else:
+            self.fail("16 random bytes shouldn't make a valid UUID4")
```
Comment (Natim) [inline syncto/tests/test_utils.py:33]: Make predictive tests, they are cases where the test can fail even if it is really unlikely.

File: syncto/tests/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    def test_bytes_are_valids_uuid4(self):
+        valid = UUID4_VALIDATOR.match(bytes_to_uuid4(os.urandom(14)))
+        if valid is None:
```
Comment (Natim) [inline syncto/tests/test_utils.py]: Just use `assert is_valid`

File: syncto/tests/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            self.fail("14 random bytes should make a valid UUID4")
+
+
+class UUID4ToBytesTest(unittest.TestCase):
+    def test_a_basic_uuid4_returns_14_bytes(self):
+        uuid = generate_made_up_uuid4()
+        uuid_bytes = uuid4_to_bytes(uuid)
+        if len(uuid_bytes) > 14:
```
Comment (Natim) [inline syncto/tests/test_utils.py]: With py.test you can use `assert` directly

File: syncto/tests/test_utils.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        uuid = generate_random_uuid4()
+        uuid_bytes = uuid4_to_bytes(uuid)
+        if len(uuid_bytes) != 16:
+            self.fail("A full random uuid4 should return 16 bytes of data. "
+                      "Returned %s" % len(uuid_bytes))
+
+
+class UUID4ConverterIdempotenceTest(unittest.TestCase):
+
+    def test_a_random_uuid4_should_return_the_same_uuid4(self):
```
Comment (Natim) [inline syncto/tests/test_utils.py:58]: Iterate on the edge case list to make sure it always works as expected.

File: syncto/utils.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+
+def base64url_encode(input, encoding='utf-8'):
+    if not isinstance(input, binary_type):
+        input = text_type(input).encode(encoding)
+    return base64.urlsafe_b64encode(input).replace(b'=', b'')
+
+
+def bytes_to_uuid4(bytes_array):
+    if len(bytes_array) > 14:
```
Comment (Natim) [inline syncto/utils.py:43]: Add comments about what you are trying to do.

== HEAVILY COMMENT HERE ==

File: syncto/views/collection.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+from pyramid.security import NO_PERMISSION_REQUIRED, forget
+
+from cliquet import Service
+from cliquet.errors import http_error, ERRORS
+from sync.client import SyncClient
+
+from syncto.utils import base64_to_uuid4
+
+
+collection = Service(name='collection',
```
Comment (Natim) [inline syncto/views/collection.py:11]: We should try to build a Cliquet backend for Kinto.

File: syncto/views/collection.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+AUTHORIZATION_HEADER = 'Authorization'
+CLIENT_STATE_HEADER = 'X-Client-State'
+
+
+@collection.get(permission=NO_PERMISSION_REQUIRED)
+def collection_get(request):
+    collection_name = request.matchdict['collection_name']
+    # Get the BID assertion
+    if AUTHORIZATION_HEADER not in request.headers or \
```
Comment (Natim) [inline syncto/views/collection.py]: Refactor in a validator.

File: syncto/views/collection.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+from cliquet import Service
+from cliquet.errors import http_error, ERRORS
+from sync.client import SyncClient
+
+from syncto.utils import base64_to_uuid4
+
+
+collection = Service(name='collection',
+                     description='Get the Firefox Sync Collection',
+                     path='/{collection_name}',
```
Comment (Natim) [inline syncto/views/collection.py]: `/buckets/syncto/collections/{collection_name}/records` ?

File: syncto/views/record.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+from cliquet import Service
+from cliquet.errors import http_error, ERRORS
+from sync.client import SyncClient
+
+from syncto.utils import base64_to_uuid4, uuid4_to_base64
+
+
+record = Service(name='record',
+                 description='Get the Firefox Sync Collection',
+                 path='/{collection_name}/{record_id}',
```
Comment (Natim) [inline syncto/views/record.py]: `/buckets/syncto/collections/{collection_name}/records/{record_id}` ?

File: syncto/views/collection.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    bid_assertion = authorization_header.split(" ", 1)[1]
+    client_state = request.headers[CLIENT_STATE_HEADER]
+    sync_client = SyncClient(bid_assertion, client_state)
+    records = sync_client.get_records(collection_name, full=True)
+
+    for r in records:
+        r['last_modified'] = int(r.pop('modified') * 1000)
+        r['id'] = base64_to_uuid4(r.pop('id'))
+
+    # Configure headers
```
Comment (Natim) [inline syncto/views/collection.py]: refactor in a helper that can be shared between collection_get and record_get

File: syncto/tests/test_functional.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+from cliquet.errors import ERRORS
+from cliquet.tests.support import FormattedErrorMixin
+
+from syncto import AUTHORIZATION_HEADER, CLIENT_STATE_HEADER
+
+from .support import BaseWebTest, unittest
+
+
+COLLECTION_URL = "/buckets/syncto/collections/{collection_id}/records"
```
Comment (almet) [inline syncto/tests/test_functional.py]: Should we name the bucket "sync" ?

File: syncto/tests/test_utils.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    bytes_to_uuid4, UUID4_VALIDATOR, uuid4_to_bytes, base64url_encode,
+    base64_to_uuid4, uuid4_to_base64, base64url_decode
+)
+
+
+def generate_random_uuid4():
+    return text_type(uuid4())
+
+
+def generate_made_up_uuid4():
```
Comment (almet) [inline syncto/tests/test_utils.py:18]: What is this used for?

File: syncto/utils.py (hunk 6 above)
Comment (almet) [inline syncto/utils.py:43]: :-)

File: syncto/tests/test_functional.py (hunk 12 above)
Comment (Natim) [inline syncto/tests/test_functional.py]: We can, the name here doesn't really matter. After discussion with Tarek, I state that they are two cases:
- When an app handle user personal data, it should use `default`
- When an app handle shared data, it is a good idea to use the app name as the bucket name.

File: syncto/views/collection.py (hunk 8 above)
Comment (almet) [inline syncto/views/collection.py]: Auto review :-) I love that.

File: syncto/views/collection.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+from cliquet import Service
+from cliquet.errors import http_error, ERRORS, raise_invalid
+from sync.client import SyncClient
+
+from syncto import AUTHORIZATION_HEADER, CLIENT_STATE_HEADER
+from syncto.utils import base64_to_uuid4, uuid4_to_base64
+
+
+collection = Service(name='collection',
+                     description='Get the Firefox Sync Collection',
```
Comment (almet) [inline syncto/views/collection.py]: This will do more than just "get" in the future, so we probably shouldn't use "get" in the description.

File: syncto/views/collection.py (hunk 8 above)
Comment (Natim) [inline syncto/views/collection.py]: Actually it was during a peer review with @n1k0 where I was writting comments @n1k0 and I talked about.

File: syncto/authentication.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+from sync.client import SyncClient
+
+from syncto import AUTHORIZATION_HEADER, CLIENT_STATE_HEADER
+
+
+def build_sync_client(request):
+    # Get the BID assertion
+    if AUTHORIZATION_HEADER not in request.headers or \
+       not request.headers[AUTHORIZATION_HEADER].lower() \
+                                                .startswith("browserid"):
```
Comment (leplatrem) [inline syncto/authentication.py]: nit: use intermediary variables (like `is_authorization_defined` `starts_with_browserid`...)

File: syncto/authentication.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+        response = http_error(httpexceptions.HTTPUnauthorized(),
+                              errno=ERRORS.MISSING_AUTH_TOKEN,
+                              message=error_msg)
+        response.headers.extend(forget(request))
+        raise response
+
+    if CLIENT_STATE_HEADER not in request.headers:
+        error_msg = "Provide the tokenserver %s header." % (
+            CLIENT_STATE_HEADER
+        )
```
Comment (leplatrem) [inline syncto/authentication.py]: nit: formatting not consistent with previous error_msg

File: syncto/tests/test_functional.py [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+
+    def __init__(self, *args, **kwargs):
+        super(FunctionalTest, self).__init__(*args, **kwargs)
+        self.collection_url = COLLECTION_URL.format(collection_id="tabs")
+        self.record_url = RECORD_URL.format(collection_id="tabs",
+                                            record_id=uuid4())
+
+    def test_authorization_header_is_required_for_collection(self):
+        resp = self.app.get(
+            self.collection_url.format(collection_id="tabs"),
```
Comment (leplatrem) [inline syncto/tests/test_functional.py]: nit: maybe you can store collection_url and record_url once, since it is always used with `'tabs'` in the test suite

File: syncto/tests/test_functional.py [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+        resp = self.app.get(self.collection_url,
+                            headers=self.headers, status=200)
+        self.assertIn('Access-Control-Allow-Origin', resp.headers)
+
+    def test_collection_handle_since_parameter(self):
+        self.app.get(self.collection_url,
+                     params={'_since': '14377478425700',
+                             '_sort': 'newest'},
+                     headers=self.headers, status=200)
+        # XXX: We should check that the sync_client is called with the
```
Comment (leplatrem) [inline syncto/tests/test_functional.py]: nit: in order to remember that this test is yet to implement, you can decorate with `@unittest.skip("todo test sync client")`

File: syncto/views/collection.py [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+        elif request.GET['_sort'] in ('-sortindex', 'index'):
+            params['sort'] = 'index'
+
+        else:
+            raise_invalid(request,
+                          location="querystring",
+                          name="_sort",
+                          description=("_sort should be one of ("
+                                       "'-last_modified', 'newest', "
+                                       "'-sortindex', 'index')"))
```
Comment (leplatrem) [inline syncto/views/collection.py]: nit: use intermediary variable (`error_msg`)

File: syncto/utils.py [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+
+
+def bytes_to_uuid4(bytes_array):
+    # If we have more that 14 bytes it means that we should provide a
+    # valid uuid4
+    if len(bytes_array) > 14:
+        uuid = text_type(UUID(hexlify(bytes_array).decode('utf-8')))
+        if UUID4_VALIDATOR.match(uuid) is None:
+            raise ValueError('Your bytes array cannot be '
+                             'converted into an UUID4.')
```
Comment (leplatrem) [inline syncto/utils.py]: nit: use intermediary variable

File: syncto/utils.py [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+            bytes_array = bytes_array.ljust(16, b'\x00')
+    # If we have less than 14 bytes, we can add two bytes that will
+    # fake a valid UUID4
+    else:
+        bytes_array = bytes_array.ljust(14, b'\x00')
+        bytes_array = (bytes_array[:6] +
+                       b'\x40' + bytes_array[6:7] +
+                       b'\x80' + bytes_array[7:14])
+
+    return text_type(UUID(hexlify(bytes_array).decode('utf-8')))
```
Comment (leplatrem) [inline syncto/utils.py]: This is bit cryptic. Please use intermediary variable (for hexlify result ?) or add comment

File: syncto/utils.py [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+
+
+def uuid4_to_bytes(uuid4):
+    bytes_array = UUID(uuid4).bytes
+    if (bytes_array[6] in (b'\x40', ord(b'\x40')) and
+            bytes_array[8] in (b'\x80', ord(b'\x80'))):
```
Comment (leplatrem) [inline syncto/utils.py]: nit: use intermediary variable with intelligible name (what does it mean sixth byte with \x40 ?

**Comment by leplatrem** [comment]
Appart from the utils.py module which is a bit scary, and would deserve some intermediary vars/docstrings/explanations/comments. I give my r+! The rest are mainly formatting nits.

GG !
